### ISSUE TITLE
Stop passing cubeb context/stream into user callbacks?

### ISSUE THREAD

**Issue by kinetiknz** [kinetiknz/cubeb#518]
All of the user callbacks accepted by libcubeb (except cubeb_device_changed_callback and cubeb_log_callback) pass a cubeb* or cubeb_stream* (depending on the context) to the user callback.  

The AudioIPC remoting code currently passes a nullptr in these situations, since none of the users inside Gecko make use of the pointer.  We also strongly discourage (and enforce, in AudioIPC) re-entering libcubeb from within a user callback, making the pointer fairly useless.

Unless there's a good reason to retain them, I'd prefer to remove these pointers from the user callback API.  That helps make it clear that re-entering libcubeb from within callbacks is discouraged and simplifies AudioIPC's implementation (as well as drops noise from the callback signatures where the pointer is ignored).

In the event there are cases where the user callback does need access to the context/stream pointer, it can access them via the user-supplied user_ptr anyway.

**Comment by padenot** [comment]
Yes, it's not particularly useful, we only ever use `user_ptr` as you say.

**Comment by achronop** [comment]
Yeah, we can remove that. We can use the `user_ptr` when we need it and it is better that way because we can handle better the thread access or the lifetime of the context/stream.

The only way I have used it in the past in the past is to differentiate callbacks for different streams but this is something we can definitely do with `user_ptr`.

**Comment by ChunMinChang** [comment]
I would love to see the context/stream pointer is removed from the arguments of the callback. It causes a twice-borrowing issue in the rust version.

**Comment by kinetiknz** [comment]
Thanks for the feedback.  I think we can go ahead with this.  My plan is to mark (or document) these arguments as deprecated, do a final libcubeb release of the current major version, then bump the major library version and remove them.
